fix: support external RTSP servers using RtspServerMode

marcpfuller opened the pull request:
<!-- Expected Commit Message Description (imported automatically by GitHub) -->
<!-- Must conform to [conventional commits guidelines](https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md) -->
<!-- Expected Commit message must contain Closes/Fixes #IssueNumber statement when there is a related issue -->

<!-- Add additional detailed description of need for change if no related issue -->

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?) n/a
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [x] I have opened a PR for the related docs change (if not, why?) 
https://github.com/edgexfoundry/edgex-docs/pull/1203

## Testing Instructions
1. run `make test`
2. verify all tests pass
3. start up edgex via edgex-compose with no-secty: `make run no-secty`
4. run device-usb-camera with `disableRtspServer` set to false `./device-usb-camera -cp -d -o`
5. verify RTSP server starts
6. run device-usb-camera with `disableRtspServer` set to true `./device-usb-camera -cp -d -o`
7. verify RTSP server does not start


## New Dependency Instructions (If applicable)
<!-- Please follow [vetting instructions](https://wiki.edgexfoundry.org/display/FA/Vetting+Process+for+3rd+Party+Dependencies) and place results here -->

Commit:
fix: Rename EnableRtspServer to DisableRtspServer

Signed-off-by: Marc-Philippe Fuller <marc-philippe.fuller@intel.com>

lenny-goodell reviewed (dismissed):
LGTM

farshidtz reviewed (dismissed):
Looks good.

Note that the snap tests are still failing because the RTCP binary path has been hardcoded. See https://github.com/edgexfoundry/device-usb-camera/pull/268#discussion_r1302646697

Commit:
fix: Move rtsp exe path to config

Signed-off-by: Marc-Philippe Fuller <marc-philippe.fuller@intel.com>

lenny-goodell commented on internal/driver/driver.go:
```diff
@@ -134,22 +141,24 @@ func (d *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
 
 	// check to see if rtsp-simple-server file/binary exists
-	_, err = os.Stat(RtspServerCmd)
+	rtspExecutable := d.ds.DriverConfigs()[RtspServerExe]
```
Should handle not set in a better way, like check for empty and return error that it isn't set

presatish commented on internal/driver/driver.go:
```diff
@@ -97,11 +97,18 @@ func (d *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
-	d.enableRtspServer, err = strconv.ParseBool(d.ds.DriverConfigs()[EnableRtspServer])
-	if err != nil {
-		return fmt.Errorf("failed to parse EnableRtspServer config: %s", err.Error())
+	// if DisableRtspServer config id blank disableRtspServer should be false
```
```suggestion
	// if DisableRtspServer config parameter is empty then it should be set to false
```

presatish commented on internal/driver/driver.go:
```diff
@@ -97,11 +97,18 @@ func (d *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
+	} else {
+		d.disableRtspServer, err = strconv.ParseBool(d.ds.DriverConfigs()[DisableRtspServer])
+		if err != nil {
+			return fmt.Errorf("failed to parse EnableRtspServer config: %s", err.Error())
```
```suggestion
			return fmt.Errorf("failed to parse DisableRtspServer config: %s", err.Error())
```

Commit:
fix: Resolve all comments made by Preethi

Signed-off-by: Marc-Philippe Fuller <marc-philippe.fuller@intel.com>

ajcasagrande commented on internal/driver/driver.go:
```diff
@@ -97,11 +97,18 @@ func (d *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
-		return fmt.Errorf("failed to parse EnableRtspServer config: %s", err.Error())
+	// if DisableRtspServer config parameter is empty then it should be set to false
+	disableRtspServer := d.ds.DriverConfigs()[DisableRtspServer]
+	if len(disableRtspServer) == 0 {
```
```suggestion
	if disableRtspServer == "" {
```

ajcasagrande commented on internal/driver/driver.go:
```diff
@@ -97,11 +97,18 @@ func (d *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
-	d.enableRtspServer, err = strconv.ParseBool(d.ds.DriverConfigs()[EnableRtspServer])
-	if err != nil {
-		return fmt.Errorf("failed to parse EnableRtspServer config: %s", err.Error())
+	// if DisableRtspServer config parameter is empty then it should be set to false
```
```suggestion
	// If DisableRtspServer config parameter is empty, then it should be
	// set to false to retain backwards compatibility
```

ajcasagrande commented on internal/driver/driver.go:
```diff
@@ -134,22 +141,28 @@ func (d *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
 	// check to see if rtsp-simple-server file/binary exists
-	_, err = os.Stat(RtspServerCmd)
+	rtspExecutable := d.ds.DriverConfigs()[RtspServerExe]
+	if len(rtspExecutable) == 0 {
```
```suggestion
	if rtspExecutable == "" {
```

ajcasagrande reviewed (changes requested):
Unfortunately there are still some issues regarding the architectural design behind this story.

When `DisableRtspServer` is set to true, then all REST APIs related to streaming will return errors (as designed). However, this does not account for the case where the user wants to disable the internal rtsp server because they want to run it separately (via docker, snap, on a different machine, etc).

In those cases, the user would (optionally) set the `RtspServerHostName` to something different, and disable the internal rtsp server.  I am not sure the best way to handle re-enabling the REST APIs.

Some options:
- Add another config var to specify that an external one is in use
- Change `DisableRtspServer` to be an enum `RtspServerMode` with:
  - `internal` (or `enabled`) - run server and enable apis
  - `external` - do not run server, but enable apis
  - `disabled` - do not run server and disable apis

Open to recommendations @farshidtz @lenny-intel

Commit:
fix: Addressed comments made by anthony

Signed-off-by: Marc-Philippe Fuller <marc-philippe.fuller@intel.com>

ajcasagrande committed:
Merge remote-tracking branch 'edgex/main' into backwards_compat

Signed-off-by: Anthony Casagrande <anthony.j.casagrande@intel.com>

ajcasagrande committed:
fix: change DisableRtspServer to RtspServerMode to support external

RtspServerMode supports values of:

internal - run RTSP server, auth server, and enable apis
external - do not run RTSP server, but enable apis and auth server
none - do not run RTSP server or auth server and disable apis
"" - Empty string defaults to internal

Signed-off-by: Anthony Casagrande <anthony.j.casagrande@intel.com>

ajcasagrande committed:
fix: rtsp auth server router not initialized properly

and revert autoStreaming to false to maintain backwards compatibility

Signed-off-by: Anthony Casagrande <anthony.j.casagrande@intel.com>

ajcasagrande committed:
Merge branch 'fix-echo-router' into backwards_compat

Signed-off-by: Anthony Casagrande <anthony.j.casagrande@intel.com>

ajcasagrande committed:
Merge remote-tracking branch 'edgex/main' into backwards_compat

Signed-off-by: Anthony Casagrande <anthony.j.casagrande@intel.com>

seanohair22 reviewed (approved):
LGTM! Tested and works with valid values and catches bad config values!

presatish reviewed (approved):
LGTM
